index.js

```diff
@@ -212,54 +240,62 @@ export function pMapIterable(
 					// Spawn if still below concurrency and backpressure limit
 					trySpawn();
 
-					try {
-						const returnValue = await mapper(await value, index++);
+					const returnValue = await mapper(await value, index++);
 
-						runningMappersCount--;
+					runningMappersCount--;
 
-						if (returnValue === pMapSkip) {
-							const index = promises.indexOf(promise);
-
-							if (index > 0) {
-								promises.splice(index, 1);
-							}
-						}
+					if (returnValue === pMapSkip) {
+						popRandomPromise(promise);
+					}
 
-						// Spawn if still below backpressure limit and just dropped below concurrency limit
-						trySpawn();
+					// Spawn if still below backpressure limit and just dropped below concurrency limit
+					trySpawn();
 
-						return {done: false, value: returnValue};
-					} catch (error) {
+					return {done: false, value: returnValue};
+				})()
+					.catch(error => {
 						isDone = true;
 						return {error};
-					}
-				})();
+					})
+					// Include a reference to the promise so `popRandomPromise` can find `indexOf`
+					.then(result => ({result, promise}));
 
 				promises.push(promise);
+				promise.then(p => {
```

Comment:
Use try/catch/async/await here. Also, would be better to use deferred promise? https://github.com/sindresorhus/p-defer